datasets/flwr_datasets/common/telemetry.py

```diff
@@ -0,0 +1,222 @@
+    if not source_file.exists():
+        try:
+            source_file.touch(exist_ok=True)
+            source_file.write_text(str(uuid.uuid4()), encoding="utf-8")
+        except PermissionError:
+            return source_id
+
+    source_id = source_file.read_text(encoding="utf-8").strip()
+
+    try:
+        uuid.UUID(source_id)
+    except ValueError:
+        source_id = "invalid"
+
+    return source_id
+
+
+# Using str as first base type to make it JSON serializable as
+# otherwise the following exception will be thrown when serializing
+# the event dict:
+# TypeError: Object of type EventType is not JSON serializable
+class EventType(str, Enum):
+    """Types of telemetry events."""
+
+    # This method combined with auto() will set the property value to
+    # the property name e.g.
+    # `START_CLIENT = auto()` becomes `START_CLIENT = "START_CLIENT"`
+    # The type signature is not compatible with mypy, pylint and flake8
+    # so each of those needs to be disabled for this line.
+    # pylint: disable-next=no-self-argument,arguments-differ,line-too-long
+    def _generate_next_value_(name: str, start: int, count: int, last_values: List[Any]) -> Any:  # type: ignore # noqa: E501
+        return name
+
+    PING = auto()
+
+    FEDERATED_DATASET_CREATED = auto()
+    PARTITIONER_CREATED = auto()
+
+
+# Use the ThreadPoolExecutor with max_workers=1 to have a queue
+# and also ensure that telemetry calls are not blocking.
+state: Dict[str, Union[Optional[str], Optional[ThreadPoolExecutor]]] = {
+    # Will be assigned ThreadPoolExecutor(max_workers=1)
+    # in event() the first time it's required
+    "executor": None,
+    "source": None,
+    "cluster": None,
+}
+
+
+# In Python 3.7 pylint will throw an error stating that
+# "Value 'Future' is unsubscriptable".
+# This pylint disable line can be remove when dropping support
+# for Python 3.7
+# pylint: disable-next=unsubscriptable-object
+def event(
+    event_type: EventType,
+    event_details: Optional[Dict[str, Any]] = None,
+) -> Future:  # type: ignore
+    """Submit create_event to ThreadPoolExecutor to avoid blocking."""
+    if state["executor"] is None:
+        state["executor"] = ThreadPoolExecutor(max_workers=1)
+
+    executor: ThreadPoolExecutor = cast(ThreadPoolExecutor, state["executor"])
+
+    result = executor.submit(create_event, event_type, event_details)
+    return result
+
+
+def create_event(event_type: EventType, event_details: Optional[Dict[str, Any]]) -> str:
+    """Create telemetry event."""
+    if state["source"] is None:
+        state["source"] = _get_source_id()
+
+    if state["cluster"] is None:
+        state["cluster"] = str(uuid.uuid4())
+
+    if event_details is None:
+        event_details = {}
+
+    date = datetime.datetime.now(tz=datetime.timezone.utc).isoformat()
+    context = {
+        "source": state["source"],
+        "cluster": state["cluster"],
+        "date": date,
+        "flower": {
```

Comment:
Note, this is changed now to `flower` to keep it consistent with the core framework.

Reply:
Let's change it to 
```suggestion
        "flwr-datasets": {
```
The `package_name` field is only there to distinguish usage of `flwr` and `flwr-nightly`. For a different project, it would be better to use a different top-level key altogether.